go/vt/vtgate/planbuilder/select2.go

```diff
@@ -0,0 +1,200 @@
+// entry.
+func getTablePlan(tableName *sqlparser.TableName, schema *Schema) (*Route, *Table, error) {
+	if tableName.Qualifier != "" {
+		return nil, nil, errors.New("tablename qualifier not allowed")
+	}
+	table, reason := schema.FindTable(string(tableName.Name))
+	if reason != "" {
+		return nil, nil, errors.New(reason)
+	}
+	if table.Keyspace.Sharded {
+		return &Route{
+			PlanID:   SelectScatter,
+			Keyspace: table.Keyspace,
+		}, table, nil
+	}
+	return &Route{
+		PlanID:   SelectUnsharded,
+		Keyspace: table.Keyspace,
+	}, table, nil
+}
+
+// processJoin produces a PlanBuilder subtree and SymbolTable for the given Join. If the left and
+// right node can be merged into one chunk, then it's a RouteBuilder. Otherwise, it's a JoinBuilder.
+func processJoin(join *sqlparser.JoinTableExpr, schema *Schema) (PlanBuilder, *SymbolTable, error) {
+	switch join.Join {
+	case sqlparser.JoinStr, sqlparser.StraightJoinStr, sqlparser.LeftJoinStr:
+	default:
+		return nil, nil, errors.New("unsupported join")
+	}
+	lplanBuilder, lsymbols, err := processTableExpr(join.LeftExpr, schema)
+	if err != nil {
+		return nil, nil, err
+	}
+	rplanBuilder, rsymbols, err := processTableExpr(join.RightExpr, schema)
+	if err != nil {
+		return nil, nil, err
+	}
+	switch lplanBuilder := lplanBuilder.(type) {
+	case *JoinBuilder:
+		return makeChunk(lplanBuilder, lsymbols, rplanBuilder, rsymbols, join, schema)
+	case *RouteBuilder:
+		switch rplanBuilder := rplanBuilder.(type) {
+		case *JoinBuilder:
+			return makeChunk(lplanBuilder, lsymbols, rplanBuilder, rsymbols, join, schema)
+		case *RouteBuilder:
+			return joinRoutes(lplanBuilder, lsymbols, rplanBuilder, rsymbols, join, schema)
+		}
+	}
+	panic("unreachable")
+}
+
+// makeChink creates a new chunk for the join operation. This function is called when
+// two chunks cannot be merged into one.
+func makeChunk(lplanBuilder PlanBuilder, lsymbols *SymbolTable, rplanBuilder PlanBuilder, rsymbols *SymbolTable, join *sqlparser.JoinTableExpr, schema *Schema) (PlanBuilder, *SymbolTable, error) {
+	err := lsymbols.AddChunk(rsymbols)
+	if err != nil {
+		return nil, nil, err
+	}
+	isLeft := false
+	if join.Join == sqlparser.LeftJoinStr {
+		isLeft = true
+	}
+	return &JoinBuilder{
+		IsLeft: isLeft,
+		Left:   lplanBuilder,
+		Right:  rplanBuilder,
+	}, lsymbols, nil
+}
+
+// joinRoutes attempts to join to RouteBuilder objects into one. If it's not possible,
+// it produces a joined RouteBuilder. Otherwise, it's a JoinBuilder.
+func joinRoutes(lplanBuilder *RouteBuilder, lsymbols *SymbolTable, rplanBuilder *RouteBuilder, rsymbols *SymbolTable, join *sqlparser.JoinTableExpr, schema *Schema) (PlanBuilder, *SymbolTable, error) {
+	switch lplanBuilder.Route.PlanID {
+	case SelectUnsharded:
+		switch rplanBuilder.Route.PlanID {
+		case SelectUnsharded:
+			// Two Routes from the same unsharded keyspace can be merged.
+			if lplanBuilder.Route.Keyspace.Name == rplanBuilder.Route.Keyspace.Name {
```

Comment:
Are you planning to also join routes here that are routed to the same shard? Or are you not going to do vindex lookups at this stage?